scripts/audio/AudioEngine.cs

```diff
@@ -0,0 +1,285 @@
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Godot;
+using TeicsoftSpectacleCards.scripts.customresource;
+
+namespace TeicsoftSpectacleCards.scripts.audio;
+
+public partial class AudioEngine : Node
+{
+    //preloaded audio to avoid stuttering when playing audio for the first time
+    private Dictionary<string, AudioStream> _preLoadedMusic = new Dictionary<string, AudioStream>();
+    private Dictionary<string, AudioStream> _preLoadedSoundFx = new Dictionary<string, AudioStream>();
+    private Dictionary<string, AudioStream> _preLoadedVoiceLines = new Dictionary<string, AudioStream>();
+
+    // two channels for music, to allow for cross fading
+    private AudioStreamPlayer _musicPlayer1;
+    private AudioStreamPlayer _musicPlayer2;
+
+    // two channels for sound effects, to allow for multiple sounds at once
+    private AudioStreamPlayer _soundFxPlayer1;
+    private AudioStreamPlayer _soundFxPlayer2;
+
+    // one channel for voice acting as there should only be one voice line at a time for clarity
+    private AudioStreamPlayer _voiceLinePlayer;
+
+    // Called when the node enters the scene tree for the first time.
+    public override void _Ready()
+    {
+        _musicPlayer1 = GetNode<AudioStreamPlayer>("MusicPlayer1");
+        _musicPlayer2 = GetNode<AudioStreamPlayer>("MusicPlayer2");
+        _soundFxPlayer1 = GetNode<AudioStreamPlayer>("SoundFxPlayer1");
+        _soundFxPlayer2 = GetNode<AudioStreamPlayer>("SoundFxPlayer2");
+        _voiceLinePlayer = GetNode<AudioStreamPlayer>("VoicePlayer");
+    }
+
+    // Music Methods //
+    public void PlayMusic(string musicFileName)
+    {
+        AudioStream audioStream = GetAudioStream(musicFileName, AudioType.Music);
+
+        if (!_musicPlayer1.Playing && !_musicPlayer2.Playing) // if nothing is playing, play on channel 1
+        {
+            _musicPlayer1.Stream = audioStream;
+            _musicPlayer1.VolumeDb = 0;
+            _musicPlayer1.Play();
+        }
+
+        else if (!_musicPlayer1.Playing) // if channel 1 is not playing, but 2 is, cross fade to channel 1
+        {
+            _musicPlayer1.Stream = audioStream;
+            _ = FadeInTrack(_musicPlayer1); // assigning Task to discard to make rider happy, as I'm not awaiting async
+            if (_musicPlayer2.Playing)
+            {
+                _ = FadeOutTrack(_musicPlayer2);
+            }
```

Comment:
This looks like a redundant check, Player 2 would have to be playing here after the check on like 41, unless I'm missing something.

Reply:
You are correct, will remove this.